fix(builtin): link module_name to directories recursively to avoid directory clashes

jbedard opened the pull request:
Fixes #1411

The overall change is to build a recursive hierarchy instead of just an array of arrays (one level of the hierarchy?). That hierarchy is then the order changes must be processed in, siblings in the hierarchy can be processed concurrently as long as the parent was already processed.

IDK if such a drastic change was really necessary, but it's the path I started down. Maybe we can just adjust the existing implementation but keep the tests?

This also does a bit less `mkdirp` calls, the `@foo/c/c/c/c` test now has I think 17 mkdirp/symlink calls instead of 21 🤷‍♂ 

~~2 of the tests fail before the change, I might try finding more cases though...~~
5 of the tests fail before the change, I think all of which are ~unique situations. I've outlined these more in the issue.

gregmagolan commented:
Nice! Thanks for taking a look at this @jbedard . I'll TAL on Friday after BazelCon.

alexeagle commented on internal/linker/test/link_node_modules.spec.ts:
```diff
@@ -150,55 +310,361 @@ describe('link_node_modules', () => {
         .toEqual('/*@foo/d/bar/fum/far*/exports = {}');
   });
 
+  it('should handle first-party packages with sibling directories', async () => {
```
These tests talking to the real FS are slower. It seems like you add all the same coverage below as in the IN/OUT unit tests above. Do we really need both? Do these new tests below add coverage that we don't get from the ones above?

jbedard replied:
They go one step further and test the use of that OUT tree, concurrently traversing it and applying it to the FS.

They could also be run with both the old and new code so I could assure I'm actually fixing things, the test failed before and passes now etc.

We could potentially remove a few now that we're (I hope) more confident everything is working. Only keep the ones that concentrate more on applying the tree to the FS...

jbedard commented:
I've updated this with a couple changes:
* rebased
* [updated some comments](https://github.com/bazelbuild/rules_nodejs/pull/1432/commits/4782d9faf0c0238f7c6f4e6caa97ed1ca0ae5649)
* [added some `reduceLink` tests](https://github.com/bazelbuild/rules_nodejs/pull/1432/commits/16b3d555be163b3cd6ea7c47dd9f3e266a1d7682) for some cases previously only the fs tests covered
* [adjustd the fs tests](https://github.com/bazelbuild/rules_nodejs/pull/1432/commits/e868fa044cf8b99c7b04b0fa3154b5527b79ffe4) to test the unique fs/link layouts, not the unique module_name => link conversions (that's what the IN/OUT tests are for). This includes deleting some cases which I think are unique module_name=> link cases, not unique fs/link cases. @alexeagle I think this addresses your point and reduces the number of these tests.

I've left those as fixup commits for now so you can see the changes. If we're good I or whoever can squash+merge.
